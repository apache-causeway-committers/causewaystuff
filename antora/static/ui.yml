static_files:
- favicon.ico
- schema/companion/v1/companion.json