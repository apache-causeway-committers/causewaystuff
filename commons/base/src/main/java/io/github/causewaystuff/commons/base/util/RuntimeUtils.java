/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.commons.base.util;

import java.util.concurrent.atomic.AtomicReference;

import org.apache.causeway.applib.services.factory.FactoryService;
import org.apache.causeway.applib.services.inject.ServiceInjector;
import org.apache.causeway.applib.services.repository.RepositoryService;
import org.apache.causeway.applib.services.wrapper.WrapperFactory;
import org.apache.causeway.commons.internal.context._Context;
import org.apache.causeway.commons.internal.ioc._IocContainer;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class RuntimeUtils {

    @Getter(lazy=true)
    private final FactoryService factoryService = getIocContainer().get(FactoryService.class).orElseThrow();

    @Getter(lazy=true)
    private final ServiceInjector serviceInjector = getIocContainer().get(ServiceInjector.class).orElseThrow();

    @Getter(lazy=true)
    private final RepositoryService repositoryService = getIocContainer().get(RepositoryService.class).orElseThrow();

    @Getter(lazy=true)
    private final WrapperFactory wrapperFactory = getIocContainer().get(WrapperFactory.class).orElseThrow();

    /**
     * Invalidates cached values, potentially useful in case the underlying application context was refreshed.
     */
    public void invalidate() {
        ((AtomicReference<Object>)factoryService).set(null);
        ((AtomicReference<Object>)serviceInjector).set(null);
        ((AtomicReference<Object>)repositoryService).set(null);
        ((AtomicReference<Object>)wrapperFactory).set(null);
        ((AtomicReference<Object>)iocContainer).set(null);
    }

    // -- HELPER

    @Getter(lazy=true, value = AccessLevel.PRIVATE)
    private final _IocContainer iocContainer = iocContainer();

    //TODO provide friendly error messages (requires a valid/bootstrapped Spring context; otherwise will fail)
    @SneakyThrows
    private _IocContainer iocContainer() {
        var envClass = _Context
            .loadClass("org.apache.causeway.core.config.environment.CausewaySystemEnvironment");
        var env = _Context.getElseFail(envClass);
        final _IocContainer iocContainer = (_IocContainer) envClass.getMethod("getIocContainer")
                .invoke(env);
        return iocContainer;
    }

}
