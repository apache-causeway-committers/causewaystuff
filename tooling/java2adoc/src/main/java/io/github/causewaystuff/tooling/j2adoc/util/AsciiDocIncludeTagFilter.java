/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.tooling.j2adoc.util;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.stream.Collectors;

import org.apache.causeway.commons.io.TextUtils;

public final class AsciiDocIncludeTagFilter {

    public static String read(final File source) {
        return TextUtils.readLinesFromFile(source, StandardCharsets.UTF_8).stream()
        //.filter(line->!containsIncludeTag(line))
        .filter(line->!isAllLineComment(line))
        .map(AsciiDocIncludeTagFilter::removeFootNoteReference)
        .collect(Collectors.joining("\n"));
    }

    public static void removeAdocExampleTags(final File source) {
        var fixedLines = TextUtils.readLinesFromFile(source, StandardCharsets.UTF_8)
        .filter(line->!isIncludeTagComment(line))
        .map(AsciiDocIncludeTagFilter::removeFootNoteReference);

        TextUtils.writeLinesToFile(fixedLines, source, StandardCharsets.UTF_8);
    }

    // -- HELPER

    private static boolean isIncludeTagComment(String line) {
        line = line.trim();
        if(!line.startsWith("//")) {
            return false;
        }
        return line.contains(" tag::")
                || line.contains(" end::");
    }

    private static boolean isAllLineComment(final String line) {
        return line.trim().startsWith("//");
    }

    private static String removeFootNoteReference(final String line) {
        if(!line.contains("// <")) {
            return line;
        }
        return line.replace("// <.>", "")
                .replace("// <1>", "")
                .replace("// <2>", "")
                .replace("// <3>", "")
                .replace("// <4>", "")
                .replace("// <5>", "")
                .replace("// <6>", "")
                .replace("// <7>", "")
                .replace("// <8>", "")
                .replace("// <9>", "")
                .replace("// <10>", "")
                .replace("// <11>", "")
                .replace("// <12>", "")
                .replace("// <13>", "")
                .replace("// <14>", "")
                .replace("// <15>", "")
                .replace("// <16>", "")
                .stripTrailing();
    }

}
