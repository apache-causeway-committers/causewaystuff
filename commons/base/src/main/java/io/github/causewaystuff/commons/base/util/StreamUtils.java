/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.github.causewaystuff.commons.base.util;

import java.util.Iterator;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;

@UtilityClass
public class StreamUtils {

    public <T> Stream<T> toStream(final Iterator<T> sourceIterator) {
        return toStream(sourceIterator, false);
    }

    public <T> Stream<T> toStream(final Iterator<T> sourceIterator, final boolean parallel) {
        Iterable<T> iterable = () -> sourceIterator;
        return StreamSupport.stream(iterable.spliterator(), parallel);
    }

    @FunctionalInterface
    public static interface Collector<T> {
        void collect(T t);
        default void terminate() {
            collect(null); // terminal
        }
    }

    /**
     * Creates a {@link Stream} from a producer that supports the visitor pattern.
     * <p>
     * A null element from the producer terminates the sequence.
     * <p>
     * Limitation: an element once produced must not be modified by the producer later,
     *      as 2 threads are racing for access to the element
     * <p>
     * Usage:
     * <pre>
     * StreamUtils.toStream(collector->{
     *      someProducer.visit(collector::collect);
     * });
     * </pre>
     * @see Collector
     */
    @SneakyThrows
    public <T> Stream<T> toStream(final Consumer<Collector<T>> consumer) {
        final var collector = new BlockingCollector<T>(
                new LinkedBlockingQueue<>(1),
                new AtomicBoolean(false));
        var producer = Thread.startVirtualThread(()->{
            consumer.accept(collector);
            collector.terminate();
        });

        return Stream.generate(collector::take)
                .takeWhile(t->{
                    var terminated = t==null;
                    if(terminated) {
                        try {
                            producer.join();
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                    return !terminated;
                });
    }

    // -- HELPER

    private record BlockingCollector<T>(
            LinkedBlockingQueue<T> q,
            AtomicBoolean terminated)
    implements Collector<T> {
        @SneakyThrows
        @Override
        public void collect(final T t) {
            if(t==null) {
                terminated.set(true);
                return;
            }
            q.put(t);
        }
        /**
         * Try to poll until q is empty or terminated.
         */
        @SneakyThrows
        public T take() {
            while(hasPotentiallyMore()) {
                var t = q.poll(10, TimeUnit.MILLISECONDS);
                if(t!=null) {
                    return t;
                }
            }
            return null;
        }

        // -- HELPER

        private boolean hasPotentiallyMore() {
            if(terminated.get()) {
                return !q.isEmpty();
            }
            return true;
        }

    }

}
