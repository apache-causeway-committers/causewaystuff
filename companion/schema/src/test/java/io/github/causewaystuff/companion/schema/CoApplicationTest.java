package io.github.causewaystuff.companion.schema;

import java.util.List;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class CoApplicationTest {
    
    CoApplication app;
    
    @BeforeEach
    void setup(){
        this.app = new CoApplication(
            "io.example.petclinic", "petclinic", "1.0.0-SNAPSHOT",
            "Petclinic", "Petclinic sample application.",
            "io.example.petclinic",
            Persistence.JPA, 
            List.of(
                new CoModule("petowner", "Pet Owner Module", "petclinic's pet owner module", List.of(
                    new CoEntity("dom.petowner", "Pet", "an individual pet, known by the Petclinic",
                        "id", Long.class),
                    new CoEntity("dom.petowner", "PetOwner", "an individual pet owner, known by the Petclinic",
                        "id", Long.class)
                    )),
                new CoModule("visit", "Visit Module", "petclinic's visit module", List.of(
                    new CoEntity("dom.visit", "Visit", "a specivic visit to the petclinic",
                        "id", Long.class)
                    ))
            ));
    }
    
    @Test
    @UseReporter(DiffReporter.class)
    void petclinic() {
        var yaml = app.toYaml();
        System.out.printf("%s%n", yaml);
        Approvals.verify(yaml);
    }
    
    @Test
    void roundtrip() {
        var yaml = app.toYaml();
        assertEquals(
                app,
                CoApplication.fromYaml(yaml));
    }
    
}
