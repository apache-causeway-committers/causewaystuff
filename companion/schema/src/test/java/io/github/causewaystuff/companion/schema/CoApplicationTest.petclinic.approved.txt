groupId: "io.example.petclinic"
artifactId: "petclinic"
version: "1.0.0-SNAPSHOT"
name: "Petclinic"
description: "Petclinic sample application."
packageName: "io.example.petclinic"
persistence: "JPA"
modules:
- id: "petowner"
  name: "Pet Owner Module"
  description: "petclinic's pet owner module"
  entities:
  - namespace: "dom.petowner"
    name: "Pet"
    description: "an individual pet, known by the Petclinic"
    primaryKeyName: "id"
    primaryKeyClass: "java.lang.Long"
  - namespace: "dom.petowner"
    name: "PetOwner"
    description: "an individual pet owner, known by the Petclinic"
    primaryKeyName: "id"
    primaryKeyClass: "java.lang.Long"
- id: "visit"
  name: "Visit Module"
  description: "petclinic's visit module"
  entities:
  - namespace: "dom.visit"
    name: "Visit"
    description: "a specivic visit to the petclinic"
    primaryKeyName: "id"
    primaryKeyClass: "java.lang.Long"
